### PR TITLE
Fixing behavior of Replicas and SearchReplicas in canary environments

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Fixing behavior of Replicas and SearchReplicas in canary environments.

The problem was, to allow to modify the config after having the license loaded,
we need to be able to remove the desanitization of the config, my first
approach was to try to use smart desanitization, based on the value obtained,
but that works only if you don't have canary environments running at the same
time.

The new approach is to allow to disable the desanitization in the case is
needed, and only using that in the Replicas and SearchReplicas initialization.

#### Ticket Link
[MM-25149](https://mattermost.atlassian.net/browse/MM-25149)

#### Original PR
#14400

### CODE REVIEW

**Comment by lieut-data** [comment]
@jespino, I'm unfamiliar with the term `canary environments` in this context. Can you elaborate?

Also, could you point me at the original PR so I can understand the changes in context?

File: app/config.go [hunk 1]
```diff
@@ -46,11 +46,20 @@ func (a *App) EnvironmentConfig() map[string]interface{} {
 	return a.Srv().EnvironmentConfig()
 }
 
+func (s *Server) UpdateConfigWithoutDesanitize(f func(*model.Config)) {
```
Comment (jespino) [inline app/config.go:49]: This is the new way to update things without desanitization (only used by my code in the `app/server.go` file). Here is the only place we set to `true` the `Set` parameter in the config store.

File: app/server_test.go [hunk 2]
```diff
@@ -38,62 +39,55 @@ func TestStartServerSuccess(t *testing.T) {
 }
 
 func TestReadReplicaDisabledBasedOnLicense(t *testing.T) {
+	cfg := model.Config{}
```
Comment (jespino) [inline app/server_test.go:42]: I change the approach of the tests, but are doing exactly the same thing.

File: config/common.go [hunk 3]
```diff
@@ -69,7 +69,9 @@ func (cs *commonStore) set(newCfg *model.Config, allowEnvironmentOverrides bool,
 
 	// Sometimes the config is received with "fake" data in sensitive fields. Apply the real
 	// data from the existing config as necessary.
-	desanitize(oldCfg, newCfg)
+	if shouldDesanitize {
```
Comment (jespino) [inline config/common.go:72]: Here is where I disable the sanitization for the special case

File: config/utils.go [hunk 4]
```diff
@@ -36,18 +36,14 @@ func desanitize(actual, target *model.Config) {
 		*target.SqlSettings.DataSource = *actual.SqlSettings.DataSource
 	}
 
-	if len(target.SqlSettings.DataSourceReplicas) == 1 && target.SqlSettings.DataSourceReplicas[0] == model.FAKE_SETTING {
```
Comment (jespino) [inline config/utils.go:39]: The changes in this file are reverting the previous approach of sanitization/desanitization.

File: model/config.go [hunk 5]
```diff
@@ -3347,8 +3347,6 @@ func (o *Config) Sanitize() {
 	}
 
 	*o.SqlSettings.DataSource = FAKE_SETTING
-	o.SqlSettings.DataSourceReplicas = []string{FAKE_SETTING}
```
Comment (jespino) [inline model/config.go:3350]: The changes in this file are reverting the previous approach of sanitization/desanitization.

**Comment by jespino** [comment]
@lieut-data Added the PR in the description and added a bunch of comments explaining the changes.

**Comment by jespino** [comment]
@lieut-data when I said canary I mean the coexistence of different versions of the server running at the same time and dealing with the same database and same clients.

This generate certain problems because one version was expecting to sanitize the things in one way and the other was expecting to sanitize in other way, and was generating the situation where you were writing data to the database that was sanitized.

**Comment by gabrieljackson** [comment]
Just waiting for the final commit with tests passing before I do a full review, but looks good.

**Comment by jespino** [comment]
@lieut-data yes, we can talk today about it. I was fallowing the preexisting pattern (that code was already there for allow only 1 replica in TE/E0), so I followed the pattern and added tests, and the test weren't working because the sanitation (I guess the code wasn't working before either).

**Comment by jespino** [comment]
@lieut-data applied the proposed changes to the code :)

File: store/sqlstore/supplier.go [hunk 6]
```diff
@@ -1178,6 +1183,10 @@ func (ss *SqlSupplier) CheckIntegrity() <-chan store.IntegrityCheckResult {
 	return results
 }
 
+func (ss *SqlSupplier) UpdateLicense(license *model.License) {
+	ss.license = license
```
Comment (lieut-data) [inline store/sqlstore/supplier.go:1187]: Given that we read and write to this value from multiple threads, should we coordinate with a mutex and avoid the eventual race detector bug?

Comment (jespino) [reply store/sqlstore/supplier.go:1187]: Yes, I can easily add a mutex there

File: store/sqlstore/supplier.go [hunk 7]
```diff
@@ -327,6 +330,10 @@ func (ss *SqlSupplier) GetMaster() *gorp.DbMap {
 }
 
 func (ss *SqlSupplier) GetSearchReplica() *gorp.DbMap {
+	if ss.license == nil {
```
Comment (lieut-data) [inline store/sqlstore/supplier.go:333]: We need to acquire the mutex on read too, right? Otherwise the mutex isn't really guarding against dirty reads.

**Comment by agnivade** [comment]
The EE test has failed btw. It has not updated the status because of GitHub rate limits. The failure looks simple enough:

```
{"level":"error","ts":1590353725.7890427,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590353735.7891996,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590353735.7897632,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590353745.7899249,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
```

**Comment by metanerd** [comment]
Screenshot from 2020-05-27 09-21-26
The workflow didn't report the status back unfortunately that it failed. It probably also ran into the rate limit.

File: app/server_test.go [hunk 8]
```diff
@@ -42,61 +43,72 @@ func TestStartServerSuccess(t *testing.T) {
...
+		dsn = os.Getenv("TEST_DATABASE_MYSQL_DSN")
+	}
+	cfg.SqlSettings = *storetest.MakeSqlSettings(driverName)
+	if dsn != "" {
+		cfg.SqlSettings.DataSource = &dsn
+	}
+	cfg.SqlSettings.DataSourceReplicas = []string{*cfg.SqlSettings.DataSource}
+	cfg.SqlSettings.DataSourceSearchReplicas = []string{*cfg.SqlSettings.DataSource}
+
 	t.Run("Read Replicas with no License", func(t *testing.T) {
```
Comment (hanzei) [inline app/server_test.go:65]: This tests fails (https://app.circleci.com/pipelines/github/mattermost/enterprise/2617/workflows/6fb78090-61eb-4950-a4dd-cc745cb778d4/jobs/9987/tests) with:
```
=== RUN   TestReadReplicaDisabledBasedOnLicense/Read_Replicas_with_no_License
{"level":"warn","ts":1590572693.4912791,"caller":"app/server.go:210","msg":"Sentry reporting is enabled, but SENTRY_DSN is not set. Disabling reporting."}
{"level":"info","ts":1590572693.4980984,"caller":"utils/i18n.go:83","msg":"Loaded system translations","for locale":"en","from locale":"/go/src/github.com/mattermost/mattermost-server/i18n/en.json"}
{"level":"info","ts":1590572693.4982169,"caller":"app/server_app_adapters.go:58","msg":"Server is initializing..."}
{"level":"debug","ts":1590572693.4989524,"caller":"app/enterprise.go:153","msg":"Loading original SAML library"}
{"level":"debug","ts":1590572693.499073,"caller":"utils/html.go:30","msg":"Parsing server templates","templates_directory":"/tmp/testlib280504788/templates"}
{"level":"info","ts":1590572693.5017102,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572693.5026684,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572703.5028367,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572703.503457,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572713.5036135,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572713.50424,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572723.5043964,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572723.5050757,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572733.5053055,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572733.5059352,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572743.5061433,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572743.5068147,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572753.5069869,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572753.5075903,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572763.5077505,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572763.5083895,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572773.5085483,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572773.5092127,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572783.5093582,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572783.5100355,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572793.510321,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572793.5108433,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572803.510985,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572803.511759,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572813.5317965,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572813.5334811,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572823.5337648,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572823.5344796,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572833.5346403,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572833.5353296,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572843.5354831,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572843.5361638,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572853.5362914,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572853.5367975,"caller":"sqlstore/supplier.go:239","msg":"Failed to ping DB","error":"dial tcp 127.0.0.1:5432: connect: connection refused","retrying in seconds":10}
{"level":"info","ts":1590572863.536942,"caller":"sqlstore/supplier.go:227","msg":"Pinging SQL","database":"master"}
{"level":"error","ts":1590572863.5375896,"caller":"sqlstore/supplier.go:235","msg":"Failed to ping DB, server will exit.","error":"dial tcp 127.0.0.1:5432: connect: connection refused"}
FAIL	github.com/mattermost/mattermost-server/v5/app	1110.065s
?   	github.com/mattermost/mattermost-server/v5/app/layer_generators	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/manual.test_load_configuration_defaults_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/manual.test_load_configuration_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_bots_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_call_log_api_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_get_bundle_path_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_get_channels_for_team_for_user_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_get_direct_channel_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_get_profile_image_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_member_channels_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_members_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_search_channels_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_search_posts_in_team_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_search_teams_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_send_mail_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_set_profile_image_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_update_user_active_plugin	[no test files]
?   	github.com/mattermost/mattermost-server/v5/app/plugin_api_tests/test_update_user_status_plugin	[no test files]
```

**Comment by jespino** [comment]
We going to disable that test, is failing without a clear reason and we will investyit later.

**Pull request merged**